firebase-firestore/src/main/java/com/google/firebase/firestore/remote/ConnectivityMonitor.java

```diff
@@ -0,0 +1,33 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.remote;
+
+import java.util.function.Consumer;
+
+/** Interface for monitoring changes in network connectivity/reachability. */
+public interface ConnectivityMonitor {
+  enum NetworkStatus {
+    UNREACHABLE,
+    REACHABLE,
+    // TODO(rsgowman): REACHABLE_VIA_CELLULAR.
+    // Leaving this off for now, since (a) we don't need it, and (b) it's somewhat messy to
```

Comment:
If we don't detect this we end up in a state where we won't respond to the network change and can fail to detect this in a way that takes minutes for us to detect.

Does the underlying managed channel builder handle this well?

If messiness is an issue, perhaps splitting the `AndroidConnectivityMonitor` in two might be worthwhile.